### PR TITLE
feat: allow multiple eligibility policies with the same group

### PR DESCRIPTION
**Pull request by wapa-soptim**
*Issue #, if available:* #39 #196

*Description of changes:*

Added the functionality to create multiple eligibility policies with the same group.

Technical changes:
V1.1.1: eligibility policies use the groupID as an unique identifier.
Unique identifier is now an uuidv4 instead of the groupID. The newly introduced entityID behaves now as the former unique identifier and references a group used by the eligibility policy.
This implementation is a **_breaking change_** as it does not migrate any eligibility policies of an existing deployment into the new data structure.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by robbycuenot** [comment]
Hit this limitation today with user entity types, and realized the same thing applies for groups too. Would love to see this merged in a future iteration. Even if it is a breaking change, it would make the tool more useful

**Comment by robbycuenot** [comment]
I do see one issue with the approach in this PR, in that you could have overlapping policies for the same accounts / OUs if the only check is for the UUID. You'd want to prevent two policies for the same user+account combinations. I haven't looked through the code enough yet to see how this is handled when one group has a policy that overlaps with a user's directly assigned policy

**Comment by Nolan-Amouyal** [comment]
Same here, I would love to see this limitation removed.

> You'd want to prevent two policies for the same user+account combinations.

I still see a use case for having two policies with the same user(or group)+account combinations: defining different settings to different permission sets.

For instance, I want to set approval to "required" only for the administrator access, but the readonly access can be automatically granted for the same user/group on the same account.

**Comment by maunzCache** [comment]
> I do see one issue with the approach in this PR, in that you could have overlapping policies for the same accounts / OUs if the only check is for the UUID. You'd want to prevent two policies for the same user+account combinations. I haven't looked through the code enough yet to see how this is handled when one group has a policy that overlaps with a user's directly assigned policy

I'd consider some checks in the frontend when creating a request as affected. For example the calculation in src/components/Request.js::getDuration() might resolve to a wrong value if multiple policies offer the same account id for a request - if i understood the procedure correctly. Really would appreciate someone else have a look at it.

There is no real solution i can think of but have an semi-hidden "advanced settings" tab which allows the user to select a specific policy to be taken into account when requesting access. However, this would require power users to understand the policies...

File: amplify/backend/function/teamRouter/src/index.py [hunk 1]
```diff
@@ -79,24 +78,26 @@ def getEntitlements(userId, groupIds):
...
-        policy = {}
-        policy['accounts'] = entitlement['Item']['accounts']
-        
-        for ou in entitlement["Item"]["ous"]:
-            data = list_account_for_ou(ou["id"])
-            policy['accounts'].extend(data)
+
+        for entitlement in entitlements["Items"]:    
+            duration = entitlement['duration']
+            if int(duration) > maxDuration:
```
Comment (maunzCache) [inline amplify/backend/function/teamRouter/src/index.py:87]: I belive this is an error if there happens to be an entitlement which covers multiple groups or users e.g. the query has more than one result for the entity_id. If so, the maxDuration will always be the highest found in the result.

Comment (soptim-wapa) [reply amplify/backend/function/teamRouter/src/index.py:87]: We have a bug report internally. As soon as we have a working solution for this, I will share this.

**Comment by soptim-wapa** [comment]
**UPDATE**

I wanted to give an update on this pull request.
There is still no timeline for when this feature will be implemented into AWS TEAM. One reason being the breaking change nature of this pull request and the lack of not having a technical migration process implemented within this pull request.

File: amplify/backend/function/teamRouter/src/index.py (hunk 1 above)
Comment (soptim-wapa) [reply amplify/backend/function/teamRouter/src/index.py:87]: Bug was fixed and the fix committed to this pull request. Thanks for your help!

**Comment by maunzCache** [comment]
Still interesting

**Comment by mhintz-clickup** [comment]
i'd be fine with a breaking change. easy enough to re-create the policies.

**Comment by CloudNinjaDev** [comment]
This is needed please re-open this issue.

**Pull request closed without merging**